common/types.go

```diff
@@ -234,16 +237,302 @@ func init() {
 }
 
 // Address represents the 20 byte address of an Ethereum account.
-type Address [AddressLength]byte
+type InternalAddress [AddressLength]byte
+type ExternalAddress [AddressLength]byte
+
+type Address struct {
+	inner AddressData
+}
+
+type AddressData interface {
+	Bytes() []byte
+	Hash() Hash
+	Hex() string
+	String() string
+	checksumHex() []byte
+	Format(s fmt.State, c rune)
+	SetBytes(b []byte)
+	MarshalText() ([]byte, error)
+	UnmarshalText(input []byte) error
+	UnmarshalJSON(input []byte) error
+	Scan(src interface{}) error
+	Value() (driver.Value, error)
+	Location() *Location
+	IsInChainScope() bool
+}
+
+func NewAddress() Address {
+	var a Address
+	var i InternalAddress
+	a.inner = &i
+	return a
+}
+
+// IsInChainScope checks if an address is a valid account in our node's sharded address space
+func (a Address) IsInChainScope() bool {
+	if a.inner == nil || a.inner.Hash() == ZeroAddr.Hash() {
+		return true
+	}
+	return NodeLocation.ContainsAddress(a)
+}
+
+func (a *Address) SetZero() *InternalAddress {
+	inner := &InternalAddress{0x0000000000000000000000000000000000000000}
+	a.inner = inner
+	return inner
+}
+
+func (a Address) InternalAddress() (*InternalAddress, error) {
+	if a.inner == nil {
+		return &InternalAddress{}, nil
+	}
+	internal, ok := a.inner.(*InternalAddress)
+	if !ok {
+		return nil, ErrInvalidScope
+	}
+	return internal, nil
+}
+
+func (a Address) Equals(b Address) bool {
+	if a.inner == nil && b.inner == nil {
+		return true
+	} else if a.inner == nil || b.inner == nil {
+		return false
+	}
+	return a.Hash() == b.Hash()
+}
 
 // BytesToAddress returns Address with value b.
 // If b is larger than len(h), b will be cropped from the left.
 func BytesToAddress(b []byte) Address {
+	if IsInChainScope(b) {
+		var i InternalAddress
+		i.SetBytes(b)
+		return Address{&i}
+	} else {
+		var e ExternalAddress
+		e.SetBytes(b)
+		return Address{&e}
+	}
+}
+
+func Bytes20ToAddress(b [20]byte) Address {
+	return BytesToAddress(b[:])
+}
+
+func InternalToAddress(i *InternalAddress) Address {
 	var a Address
-	a.SetBytes(b)
+	a.inner = i
 	return a
 }
 
+// EncodeRLP serializes b into the Ethereum RLP block format.
+func (a Address) EncodeRLP(w io.Writer) error {
+	if a.inner == nil {
+		a.inner = &InternalAddress{}
+	}
+	return rlp.Encode(w, a.inner)
+}
+
+// DecodeRLP decodes the Ethereum
+func (a *Address) DecodeRLP(s *rlp.Stream) error {
+	temp := make([]byte, 0, 20)
+	if err := s.Decode(&temp); err != nil {
+		return err
+	}
+	*a = BytesToAddress(temp)
+	return nil
+}
+
+// SetBytes sets the address to the value of b.
+// If b is larger than len(a), b will be cropped from the left.
+func (a *Address) SetBytes(b []byte) {
```

Comment:
I don't think this function should exist. We should never be fiddling with the bytes of an address. We set the bytes once upon construction (see proposed `NewAddress(b [20]bytes)` above) and we should never modify those bytes again.

I looked through the code, and I believe everywhere you use `SetBytes()` could easily just use the constructor. Let me know if there is some use case I am overlooking

Reply:
Fixed